datafusion/sql/src/statement.rs

```diff
@@ -710,6 +710,15 @@ impl<S: ContextProvider> SqlToRel<'_, S> {
                     *statement,
                     &mut planner_context,
                 )?;
+
+                if data_types.is_empty() {
+                    for dt in plan.get_parameter_types()?.into_values().flatten() {
+                        data_types.push(dt);
+                    }
+                    data_types.sort();
+                    planner_context.with_prepare_param_data_types(data_types.clone());
+                }
+
                 Ok(LogicalPlan::Statement(PlanStatement::Prepare(Prepare {
```

Comment:
Another thing I suggest we do (as a follow on PR) is move the type checking / construction into a `LogicalPlanBuilder` method rather than entirely in the SQL planner, so that other frontends can use it too

Reply:
Sure! I can create an issue for that, but not sure what would be the scope of it. I potentially see these functions, 
https://github.com/apache/datafusion/blob/d5ecaca33c8472a6520e081ac3c9ad3dac098caf/datafusion/sql/src/planner.rs#L229

https://github.com/apache/datafusion/blob/d5ecaca33c8472a6520e081ac3c9ad3dac098caf/datafusion/sql/src/planner.rs#L520

https://github.com/apache/datafusion/blob/d5ecaca33c8472a6520e081ac3c9ad3dac098caf/datafusion/sql/src/planner.rs#L557

https://github.com/apache/datafusion/blob/d5ecaca33c8472a6520e081ac3c9ad3dac098caf/datafusion/expr/src/logical_plan/plan.rs#L1491